File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldExceptionsArray.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldExceptionsArray.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.List;
import java.util.ArrayList;

import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class HistoricElementFieldExceptionsArray extends Element
{
    private final List<HistoricElementFieldExceptions> _exceptions;

    public HistoricElementFieldExceptionsArray(List<String> badFields)
    {
        this._exceptions = new ArrayList<HistoricElementFieldExceptions>(badFields.size());
        for(int i = 0; i < badFields.size(); i++)
        {
        	this._exceptions.add(new HistoricElementFieldExceptions(badFields.get(i)));
        }
    }
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("FieldException"));
    }
	
    public Name name() throws Exception
    {
    	return new Name("fieldExceptions");
    }
    
    public int numValues()
    {
    	return this._exceptions.size();
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return true;
    }
    
    public boolean isNull()
    {
    	return false;
    }

    public Object getValue(int index)
    {
        return this._exceptions.get(index);
    }

    public Element getValueAsElement(int index)
    {
        return this._exceptions.get(index);
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s[] = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        
        for(int i = 0; i < this._exceptions.size(); i++)
        {
        	result.append(this._exceptions.get(i).prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }

}

File: BEmu_cpp/src/HistoricalDataRequest/HistoricElementFieldExceptionsArray.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest/HistoricElementFieldExceptionsArray.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "HistoricalDataRequest/HistoricElementFieldExceptionsArray.h"
#include "HistoricalDataRequest/HistoricElementFieldExceptions.h"

#include "BloombergTypes/Name.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace HistoricalDataRequest
	{
		HistoricElementFieldExceptionsArray::HistoricElementFieldExceptionsArray(const std::vector<std::string>& badFields)
		{
			for(std::vector<std::string>::const_iterator iter = badFields.begin(); iter != badFields.end(); ++iter)
			{
				std::string str = *iter;				
				boost::shared_ptr<HistoricElementFieldExceptions> elmP(new HistoricElementFieldExceptions(str));
				
				this->_exceptions.push_back(elmP);
			}
		}

		HistoricElementFieldExceptionsArray::~HistoricElementFieldExceptionsArray()
		{
			this->_exceptions.clear();
		}


		Name HistoricElementFieldExceptionsArray::name() const
		{
			Name result("fieldExceptions");
			return result;
		}

		size_t HistoricElementFieldExceptionsArray::numValues() const
		{
			return this->_exceptions.size();
		}

		size_t HistoricElementFieldExceptionsArray::numElements() const
		{
			return 0;
		}

		SchemaElementDefinition HistoricElementFieldExceptionsArray::elementDefinition() const
		{
			::blpapi_DataType_t dtype = (::blpapi_DataType_t)this->datatype();
			SchemaElementDefinition result(dtype, Name("FieldException"));
			return result;
		}

		
		bool HistoricElementFieldExceptionsArray::isNull() const
		{
			return false;
		}

		bool HistoricElementFieldExceptionsArray::isArray() const
		{
			return true;
		}

		bool HistoricElementFieldExceptionsArray::isComplexType() const
		{
			return false;
		}


		boost::shared_ptr<ElementPtr> HistoricElementFieldExceptionsArray::getValueAsElement(int index) const
		{
			return boost::dynamic_pointer_cast<ElementPtr>(this->_exceptions.at(index));
		}


		std::ostream& HistoricElementFieldExceptionsArray::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << "fieldExceptions = {" << std::endl;

			for(std::vector< boost::shared_ptr<HistoricElementFieldExceptions> >::const_iterator iter = this->_exceptions.begin(); iter != this->_exceptions.end(); ++iter)
			{
				boost::shared_ptr<HistoricElementFieldExceptions> elm = *iter;
				elm->print(stream, level + 1, spacesPerLevel);
			}

			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}